pkg/planner/core/operator/logicalop/logical_datasource.go

```diff
@@ -621,3 +633,90 @@ func preferKeyColumnFromTable(dataSource *DataSource, originColumns []*expressio
 	}
 	return resultColumn, resultColumnInfo
 }
+
+// analyzeFTSFunc checks whether FTS function is used and is a valid one.
+// Then convert the function to index call because it can not be executed without the index.
+func (ds *DataSource) analyzeFTSFunc() error {
+	if !ds.SCtx().HasFTSFunc() {
+		return nil
+	}
+	idx2FastCheck := make(map[*model.IndexInfo]intset.FastIntSet)
+	idSetForCheck := intset.NewFastIntSet()
+	for _, index := range ds.FtsIndexes {
+		s := intset.NewFastIntSet()
+		for _, col := range index.Columns {
+			s.Insert(int(ds.TableInfo.Columns[col.Offset].ID))
+		}
+		idx2FastCheck[index] = s
+	}
+	var matchedIdx *model.IndexInfo
+	var matchedFunc *expression.ScalarFunction
+	var matchedCondPos int
+	matchedColumns := make([]*expression.Column, 0, 2)
+	for i, cond := range ds.PushedDownConds {
+		sf, ok := cond.(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.FTSMatchWord {
+			if expression.ContainsFullTextSearchFn(cond) {
+				return plannererrors.ErrWrongUsage.FastGen(plannererrors.FTSWrongPlace)
+			}
+			continue
+		}
+		idSetForCheck.Clear()
+		for i := 1; i < len(sf.GetArgs()); i++ {
+			col := sf.GetArgs()[i].(*expression.Column)
```

Comment:
Where do we check to make sure all the args are `Column`s (except for the first one)?

Reply:
It's checked when build it in `func (c *ftsMatchWordFunctionClass) getFunction(ctx BuildContext, args []Expression)`